book_extractors/common/extractors/base_extractor.py

```diff
@@ -19,6 +21,31 @@ def __init__(self, key_of_cursor_location_dependent=None, options=None):
 
         self.metadata_collector = MetadataCollector()
 
+    def _build_dependencies_graph(self, dependencies_contexts):
+        if self._dependencies is not None and dependencies_contexts != {}:
+            if len(dependencies_contexts) < len(self._dependencies):
+                while len(dependencies_contexts) < len(self._dependencies):
+                    dependencies_contexts.append(None)
+
+                missing_contexts = []
+                for ext, context in zip(self._dependencies, dependencies_contexts):
+                    if context is None:
+                        missing_contexts.append(ext.__name__)
+
+                raise DependencyConfigurationException('There is a mismatch between the amount of dependencies and dependency contexts! '
+                                                       'Please explicitly enter a context for each dependency. '
+                                                       'The following extractors are missing a context: {0}'.format(missing_contexts))
+
+            new_dependencies = [(extractor, context) for extractor, context in zip(self._dependencies, dependencies_contexts)]
+            self._dependencies_graph += new_dependencies
+
+    def _set_dependencies(self, dependencies, dependencies_contexts):
+        for dependency in dependencies:
+            self._dependencies.append(dependency)
+
+        if dependencies_contexts is not None:
+            self._build_dependencies_graph(dependencies_contexts)
```

Comment:
Tässä käy nyt niin, että jos `dependencies_contexts` on `None` (joka on `dependencies_contexts` default arvo `configure_extractor` funktion ja jokaisen luokan constructorien parametreissa) niin `_dependencies_graph` ei rakennu. Tämä toimii näin sen takia, että annetaan ekaksi superluokkien lisätä omat dependencyt `_set_dependencies` kutsulla `_dependencies` muuttujaan. Näissä kutsuissa `dependencies_contexts` tulee olemaan `None`, koska superluokkien `__init__` kutsutaan aliluokan constructorista, eikä pipelinen luonnin yhteydessä `configure_extractor` kautta. Ainoastaan alin luokka kutsutaan `configure_extractor` kautta.  Kun tuo tapahtuu, sille annetaan sen ja jokaisen yliluokan kontekstit, ja kun se lisää depsinsä `_set_dependencies` kutsulla, `dependencies_contexts` ei olekaan enää `None`, ja graph rakennetaan. Tällä myös vältymme graphin luomiselta n-kertaa.